Readme.txt
EMP Biters 1.1.0
================

Version 1.1.0 was released April 26, 2017, was tested using Factorio v0.15.2, and was authored by Supercheese.

Many mods give the player a whole host of new weapons and defenses to play with, but the poor biters are left behind. It's time they received an upgrade!
Now, large biters have a chance to give off an EMP blast when they die, temporarily disabling nearby turrets, radars, and logistic & construction robots. Large spitters also have a chance to EMP with their spit.
The larger the enemy and the higher the evolution factor, the bigger the chance of EMP is.

Gun turrets and flamethrower turrets are only disabled by EMP for half as long as laser turrets, to encourage using mixed defenses.
Big Worms have a high chance to cause a large EMP blast with their spit, to make turret creep more challenging.

This mod also supports Bob's Enemies, Dytech War, Natural Evolution Enemies, and 5dim Battlefield.

The area of effect, minimum & maximum durations, etc. are stored in config.lua. Feel free to edit these values to increase or decrease the difficulty.


FAQs:
-----

Q) How are biters/spitters able to generate an EMP anyway?
A) Magic. The same magic that lets FactorioGuy carry around 20 Tanks, Locomotives, and Steam Engines in his pocket.

Q) Jeez, this mod really makes things a lot harder. How do I defend my base now?
A) Landmines. Lots of landmines. Or spread your turrets out more. Maybe even make a secondary defensive line. I'm sure you'll come up with something!
EMP attacks won't become a threat until you reach Big Biters anyway, and by that time you might already have construction robots to automatically repair damage and replace destroyed structures.

Q) Is there any way I can fix my structures when they've been hit with an EMP attack?
A) Yes, you can pick them up and replace them. You can use bots to deconstruct/reconstruct them too, but this is risky because the bots themselves are vulnerable to EMP. Other than this, you must wait for the EMP effect to wear off.

Q) Will this mod hurt my FPS/UPS?
A) I have benchmarked this mod under very heavy biter attacks, and even then it still had better performance than the popular YARM mod. So if your game runs fine while monitoring several sites with YARM, it should run just fine with this.
Of note, when no entities are affected by EMP, the mod shuts down its on_tick function to improve performance.

Q) What happens if, after I load the mod, later I chicken out and decide I want to remove it?
A) Well, if you remove this mod from a running game where some entities are disabled by EMP, those entities will not be automatically re-enabled; you'll have to manually pick them up and replace them to re-enable them.
Other than that everything else should be fine.


Credits:
--------

The EMP blast graphics were obtained from: http://opengameart.org/content/animated-blue-ring-explosion
They were uploaded by the user "Grahhhhh" under the CC-0 (Creative Commons 0) license. (See: http://creativecommons.org/publicdomain/zero/1.0/)

The clouds of sparks that appear above disabled turrets are modified from Factorio base graphics.

Thanks to the forum and #factorio IRC denizens for advice & bugtesting.


See also the associated forum thread to give feedback, view screenshots, etc.:

https://forums.factorio.com/viewtopic.php?f=93&t=22294


Changelog.txt
Version 1.0.0 - March 26, 2016
------------------------------
Now, large biters have a chance to give off an EMP blast when they die, temporarily disabling nearby turrets, radars, and logistic & construction robots. Large spitters also have a chance to EMP with their spit.
The larger the enemy and the higher the evolution factor, the bigger the chance of EMP is.


Version 1.0.1 - March 27, 2016
------------------------------
Increased default difficulty settings: all EMP attacks have increased disable duration.
To compensate, gun turrets are now disabled by EMP for only half as long as other entities. This should encourage using mixed defenses.


Version 1.0.2 - March 28, 2016
------------------------------
Added support for the 5dim Battlefield mod and its new enemies.
Fixed a crash that could happen when picking up a structure that was disabled by EMP.
Fixed Big Worms' chance to cause an EMP being lower than intended; their chance now increases from 0% at 0% evolution and reaches 100% by 50% evolution -- previously they would not reach 100% chance until 100% evo.


Version 1.0.3 - March 31, 2016
------------------------------
Fixed a crash that could happen when a radar-type entity was disabled by EMP.


Version 1.0.4 - May 4, 2016
---------------------------
Added support for Bob's Enemies 0.12.7.


Version 1.0.5 - May 6, 2016
---------------------------
Optimize mod performance by caching calls to game.tick.
Slightly reduce how often sounds are emitted by entities when they are disabled by EMP.


Version 1.0.6 - June 30, 2016
-----------------------------
Updated for Factorio 0.13.
Flamethrower turrets are now also vulnerable to EMP, but like gun turrets, they are only disabled for half as long as laser turrets.
Further performance optimizations.
Fix typo leading to possible crash when entity reference becomes invalid while disabled (thanks to caspark for catching this).


Version 1.0.7 - August 27, 2016
-------------------------------
Updated for Factorio 0.14.


Version 1.1.0 - April 26, 2017
------------------------------
Updated for Factorio 0.15.


config.lua
-- This is the lookup table for EMP biters. Rather than add its own custom biters, this mod gives certain extant biters a chance to cause an EMP blast and disable nearby turrets.
-- Only biters that have an entry in this table are eligible to give off an EMP blast on death.

-- The format for entries is: 							{size, maxChance, minDuration, maxDuration}

-- Size is in tiles. This is the "radius" (it's a square region actually) of the area of effect around the dying biter for the EMP blast.

-- MaxChance is in percent. This is the maximum chance that a biter will trigger an EMP blast when it dies.
-- The chance starts out at 0% when the evolution factor is 0 and increases up to the maximum chance as the evolution factor rises.
-- MaxChance can exceed 100, and this just means that it will reach 100% chance at a lower evo factor.
-- For instance, the Big Worm has a default value of 200, and this means that it reaches 100% chance at 50% evolution factor.

-- Durations are in ticks (60 ticks per second).
-- The exact duration of the EMP disabling is randomized between the minimum and maximum values.

EMPBiterTable = {}

-- Vanilla Biters
EMPBiterTable["big-biter"] = 							{10, 15,  300, 600}
EMPBiterTable["behemoth-biter"] = 						{15, 25,  300, 900}

-- Bob's Enemies 0.12.6 and earlier
EMPBiterTable["bob-bigger-biter"] =						{10, 20,  360, 900}
EMPBiterTable["bob-biggest-biter"] =					{12, 25,  600, 900}
EMPBiterTable["bob-fire-biter"] =						{14, 30,  600, 1200}
EMPBiterTable["bob-titan-biter"] =						{16, 35,  900, 1500}
EMPBiterTable["bob-behemoth-biter"] =					{18, 40,  1200, 2400}
-- Bob's Enemies 0.12.7 and later
EMPBiterTable["bob-big-piercing-biter"] =				{10, 20,  360, 900}
EMPBiterTable["bob-huge-acid-biter"] =					{12, 25,  600, 900}
EMPBiterTable["bob-huge-explosive-biter"] =				{12, 25,  600, 900}
EMPBiterTable["bob-giant-poison-biter"] =				{14, 30,  600, 1200}
EMPBiterTable["bob-giant-fire-biter"] =					{14, 30,  600, 1200}
EMPBiterTable["bob-titan-biter"] =						{16, 35,  900, 1500}
EMPBiterTable["bob-behemoth-biter"] =					{18, 40,  1200, 2400}
EMPBiterTable["bob-leviathan-biter"] =					{20, 50,  1500, 3000}

-- Dytech War
EMPBiterTable["teen-berserk-biter"] = 					{10, 10,  300, 600}
EMPBiterTable["teen-elder-biter"] = 					{11, 15,  300, 900}
EMPBiterTable["teen-king-biter"] = 						{12, 20,  300, 1200}
EMPBiterTable["teen-queen-biter"] = 					{13, 20,  300, 1500}
EMPBiterTable["adult-berserk-biter"] = 					{14, 20,  600, 1200}
EMPBiterTable["adult-elder-biter"] = 					{15, 25,  600, 1500}
EMPBiterTable["adult-king-biter"] = 					{20, 50,  900, 2100}
EMPBiterTable["adult-queen-biter"] = 					{25, 100, 900, 2400}

EMPBiterTable["dyzilla-unit-1"] = 						{20, 100, 1200, 1800}
EMPBiterTable["dyzilla-unit-2"] = 						{25, 100, 1200, 2100}
EMPBiterTable["dyzilla-unit-3"] = 						{40, 100, 1200, 2400}
EMPBiterTable["dyzilla-unit-4"] = 						{50, 100, 1200, 3000}

-- Natural Evolution Enemies
EMPBiterTable["big-biter-Mk2"] = 						{12, 20,  300, 600}
EMPBiterTable["big-biter-Mk3"] = 						{14, 20,  300, 900}

-- 5dim Battlefield
EMPBiterTable["5d-medium-biter-laser"] = 				{5,  20,  300, 600}
EMPBiterTable["5d-medium-biter-physical"] = 			{5,  20,  300, 600}
EMPBiterTable["5d-medium-biter-explosive"] =			{5,  20,  300, 600}
EMPBiterTable["5d-big-biter-laser"] = 					{10, 25,  600, 900}
EMPBiterTable["5d-big-biter-physical"] = 				{10, 25,  600, 900}
EMPBiterTable["5d-big-biter-explosive"] = 				{10, 25,  600, 900}


-- This is the lookup table for EMP spitters & worms. Rather than add its own custom spitters, this mod gives certain extant spitters a chance to disable a turret when they hit it with their spit.
-- Big worms have a 100% chance to cause an EMP blast, to make turret creep more challenging.

-- The format is the same as above:						{size, maxChance, minDuration, maxDuration}

EMPSpitterTable = {}

-- Vanilla Spitters
EMPSpitterTable["EMP-acid-splash-purple-big"] =			{1,  20,  300, 600}  -- also "bob-huge-acid-spitter"
EMPSpitterTable["EMP-acid-splash-purple-behemoth"] =	{2,  25,  300, 900}

-- Vanilla Big Worm
EMPSpitterTable["EMP-acid-splash-purple-worm"] =		{20, 200, 600, 1200} -- Suck it, turret creep!

-- Bob's Enemies 0.12.6 and earlier
EMPSpitterTable["EMP-acid-splash-explosive"] =			{1,  20,  300, 420}  -- "bob-bigger-spitter"/"bob-huge-explosive-spitter" and "bob-big-explosive-worm-turret"
EMPSpitterTable["EMP-acid-splash-fire"] =				{2,  20,  300, 540}  -- "bob-biggest-spitter"/"bob-giant-fire-spitter" and "bob-big-fire-worm-turret"
EMPSpitterTable["EMP-acid-splash-poison"] =				{3,  20,  420, 600}  -- "bob-poison-spitter"/"bob-giant-poison-spitter" and "bob-big-poison-worm-turret"
EMPSpitterTable["EMP-acid-splash-titan"] =				{4,  25,  600, 900}  -- "bob-titan-spitter"
EMPSpitterTable["EMP-acid-splash-behemoth"] =			{5,  25,  900, 1200} -- "bob-behemoth-spitter"
-- Bob's Enemies 0.12.7 and later
EMPSpitterTable["EMP-acid-splash-electric"] =			{1,  20,  300, 420}  -- "bob-big-electric-spitter"
EMPSpitterTable["EMP-acid-splash-leviathan"] =			{7,  50,  900, 1500} -- "bob-leviathan-spitter"

-- Dytech War
EMPSpitterTable["EMP-acid-splash-berserker"] =			{2,  25,  300, 600}  -- All Berzerker-type spitters
EMPSpitterTable["EMP-acid-splash-elder"] =				{3,  40,  600, 900}  -- All Elder-type spitters
EMPSpitterTable["EMP-acid-splash-king"] =				{5,  50,  600, 1200} -- All King-type spitters

-- Natural Evolution Enemies
EMPSpitterTable["EMP-acid-splash-infected"] =			{2,  25,  300, 720}  -- "big-spitter-Mk2"
EMPSpitterTable["EMP-acid-splash-mutated"] =			{2,  25,  300, 900}  -- "big-spitter-Mk3"


-- This table enumerates all entity-types that are vulnerable to EMP. Please don't add or remove anything from this table at this time.

vulnerableEntityTypes = {}

vulnerableEntityTypes["ammo-turret"] = 				true
vulnerableEntityTypes["fluid-turret"] = 			true
vulnerableEntityTypes["electric-turret"] = 			true
vulnerableEntityTypes["logistic-robot"] = 			true
vulnerableEntityTypes["construction-robot"] = 		true
vulnerableEntityTypes["radar"] = 					true


prototypes/entities.lua
data:extend({ -- Clouds of sparks
	{
		type = "smoke-with-trigger",
		name = "disabling-EMP-cloud-1",
		flags = {"not-on-map"},
		show_when_smoke_off = true,
		animation =
		{
			filename = "__EMP Biters__/graphics/EMP-cloud-1.png",
			priority = "high",
			width = 74,
			height = 70,
			frame_count = 16,
			scale = 0.9,
			animation_speed = 0.25
		},
		slow_down_factor = 0,
		affected_by_wind = false,
		cyclic = true,
		duration = 60 * 50,
		fade_away_duration = 2 * 60,
		spread_duration = 10,
		color = { r = 0, g = 0.5, b = 1.0, a = 0.5 },
	},
	{
		type = "smoke-with-trigger",
		name = "disabling-EMP-cloud-2",
		flags = {"not-on-map"},
		show_when_smoke_off = true,
		animation =
		{
			filename = "__EMP Biters__/graphics/EMP-cloud-2.png",
			priority = "high",
			width = 74,
			height = 70,
			frame_count = 16,
			scale = 0.9,
			animation_speed = 0.25
		},
		slow_down_factor = 0,
		affected_by_wind = false,
		cyclic = true,
		duration = 60 * 50,
		fade_away_duration = 2 * 60,
		spread_duration = 10,
		color = { r = 0, g = 0.5, b = 1.0, a = 0.5 },
	},
	{
		type = "smoke-with-trigger",
		name = "disabling-EMP-cloud-3",
		flags = {"not-on-map"},
		show_when_smoke_off = true,
		animation =
		{
			filename = "__EMP Biters__/graphics/EMP-cloud-3.png",
			priority = "high",
			width = 74,
			height = 70,
			frame_count = 16,
			scale = 0.9,
			animation_speed = 0.25
		},
		slow_down_factor = 0,
		affected_by_wind = false,
		cyclic = true,
		duration = 60 * 50,
		fade_away_duration = 2 * 60,
		spread_duration = 10,
		color = { r = 0, g = 0.5, b = 1.0, a = 0.5 },
	},
	{
		type = "smoke-with-trigger",
		name = "disabling-EMP-cloud-4",
		flags = {"not-on-map"},
		show_when_smoke_off = true,
		animation =
		{
			filename = "__EMP Biters__/graphics/EMP-cloud-4.png",
			priority = "high",
			width = 74,
			height = 70,
			frame_count = 16,
			scale = 0.9,
			animation_speed = 0.25
		},
		slow_down_factor = 0,
		affected_by_wind = false,
		cyclic = true,
		duration = 60 * 50,
		fade_away_duration = 2 * 60,
		spread_duration = 10,
		color = { r = 0, g = 0.5, b = 1.0, a = 0.5 },
	},

-- Sound effects
	{
		type = "explosion",
		name = "EMP-disabled-sound-1",
		flags = {"not-on-map"},
		animations =
		{
			{
				filename = "__EMP Biters__/graphics/null.png",
				priority = "low",
				width = 32,
				height = 32,
				frame_count = 1,
				line_length = 1,
				animation_speed = 1
			},
		},
		light = {intensity = 0, size = 0},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/Disabled-1.ogg",
				volume = 0.5
			},
		},
	},
	{
		type = "explosion",
		name = "EMP-disabled-sound-2",
		flags = {"not-on-map"},
		animations =
		{
			{
				filename = "__EMP Biters__/graphics/null.png",
				priority = "low",
				width = 32,
				height = 32,
				frame_count = 1,
				line_length = 1,
				animation_speed = 1
			},
		},
		light = {intensity = 0, size = 0},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/Disabled-2.ogg",
				volume = 0.5
			},
		},
	},
	{
		type = "explosion",
		name = "EMP-disabled-sound-3",
		flags = {"not-on-map"},
		animations =
		{
			{
				filename = "__EMP Biters__/graphics/null.png",
				priority = "low",
				width = 32,
				height = 32,
				frame_count = 1,
				line_length = 1,
				animation_speed = 1
			},
		},
		light = {intensity = 0, size = 0},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/Disabled-3.ogg",
				volume = 0.5
			},
		},
	},
	{
		type = "explosion",
		name = "EMP-disabled-sound-4",
		flags = {"not-on-map"},
		animations =
		{
			{
				filename = "__EMP Biters__/graphics/null.png",
				priority = "low",
				width = 32,
				height = 32,
				frame_count = 1,
				line_length = 1,
				animation_speed = 1
			},
		},
		light = {intensity = 0, size = 0},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/Disabled-4.ogg",
				volume = 0.5
			},
		},
	},

-- Blue EMP Blasts
	{
		type = "explosion",
		name = "EMP-blast-small",
		flags = {"not-on-map"},
		animations =
		{
			{
			filename = "__EMP Biters__/graphics/EMP-blast.png",
			priority = "high",
			width = 200,
			height = 200,
			scale = 1,
			frame_count = 20,
			line_length = 5,
			shift = {0, 0},
			animation_speed = 0.75
			},
		},
		light = {intensity = 2, size = 5},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/EMP.ogg",
				volume = 1.0
			},
		},
	},
	{
		type = "explosion",
		name = "EMP-blast-medium",
		flags = {"not-on-map"},
		animations =
		{
			{
			filename = "__EMP Biters__/graphics/EMP-blast.png",
			priority = "high",
			width = 200,
			height = 200,
			scale = 2,
			frame_count = 20,
			line_length = 5,
			shift = {0, 0},
			animation_speed = 0.75
			},
		},
		light = {intensity = 2, size = 10},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/EMP.ogg",
				volume = 1.5
			},
		},
	},
	{
		type = "explosion",
		name = "EMP-blast-large",
		flags = {"not-on-map"},
		animations =
		{
			{
			filename = "__EMP Biters__/graphics/EMP-blast.png",
			priority = "high",
			width = 200,
			height = 200,
			scale = 3,
			frame_count = 20,
			line_length = 5,
			shift = {0, 0},
			animation_speed = 0.75
			},
		},
		light = {intensity = 2, size = 15},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/EMP.ogg",
				volume = 1.5
			},
		},
	},
	{
		type = "explosion",
		name = "EMP-blast-huge",
		flags = {"not-on-map"},
		animations =
		{
			{
			filename = "__EMP Biters__/graphics/EMP-blast.png",
			priority = "high",
			width = 200,
			height = 200,
			scale = 4,
			frame_count = 20,
			line_length = 5,
			shift = {0, 0},
			animation_speed = 0.75
			},
		},
		light = {intensity = 2, size = 20},
		sound =
		{
			{
				filename = "__EMP Biters__/sound/EMP.ogg",
				volume = 1.5
			},
		},
	},
})

prototypes/projectiles.lua
-- Vanilla
local EMP_acid_projectile_big = util.table.deepcopy(data.raw["projectile"]["acid-projectile-purple"])
local EMP_acid_projectile_behemoth = util.table.deepcopy(data.raw["projectile"]["acid-projectile-purple"])
local EMP_acid_projectile_worm = util.table.deepcopy(data.raw["projectile"]["acid-projectile-purple"])

EMP_acid_projectile_big.name = "EMP-acid-projectile-big"
EMP_acid_projectile_big.action.action_delivery.target_effects[2].entity_name = "EMP-acid-splash-purple-big"
EMP_acid_projectile_big.action.action_delivery.target_effects[2].trigger_created_entity = "true"

EMP_acid_projectile_behemoth.name = "EMP-acid-projectile-behemoth"
EMP_acid_projectile_behemoth.action.action_delivery.target_effects[2].entity_name = "EMP-acid-splash-purple-behemoth"
EMP_acid_projectile_behemoth.action.action_delivery.target_effects[2].trigger_created_entity = "true"

EMP_acid_projectile_worm.name = "EMP-acid-projectile-worm"
EMP_acid_projectile_worm.action.action_delivery.target_effects[2].entity_name = "EMP-acid-splash-purple-worm"
EMP_acid_projectile_worm.action.action_delivery.target_effects[2].trigger_created_entity = "true"

local EMP_acid_splash_big = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
local EMP_acid_splash_behemoth = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
local EMP_acid_splash_worm = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])

EMP_acid_splash_big.name = "EMP-acid-splash-purple-big"
EMP_acid_splash_behemoth.name = "EMP-acid-splash-purple-behemoth"
EMP_acid_splash_worm.name = "EMP-acid-splash-purple-worm"

data:extend({EMP_acid_projectile_big, EMP_acid_projectile_behemoth, EMP_acid_projectile_worm, EMP_acid_splash_big, EMP_acid_splash_behemoth, EMP_acid_splash_worm})

data.raw["unit"]["big-spitter"].attack_parameters.ammo_type.action.action_delivery.projectile = "EMP-acid-projectile-big"
data.raw["unit"]["behemoth-spitter"].attack_parameters.ammo_type.action.action_delivery.projectile = "EMP-acid-projectile-behemoth"

if not data.raw["projectile"]["area-acid-projectile-purple"] then -- Bob's Enemies already edits the Big Worm
	data.raw["turret"]["big-worm-turret"].attack_parameters.ammo_type.action.action_delivery.projectile = "EMP-acid-projectile-worm"
end


-- Bob's Enemies

if data.raw["projectile"]["area-acid-projectile-purple"] then -- Bob's big worm
	data.raw["projectile"]["area-acid-projectile-purple"].action.action_delivery.target_effects[1] = {
		type = "create-entity",
		entity_name = "EMP-acid-splash-purple-worm",
		trigger_created_entity = "true"
	}
end

if data.raw["unit"]["bob-huge-acid-spitter"] then
	data.raw["unit"]["bob-huge-acid-spitter"].attack_parameters.ammo_type = {
        category = "biological",
        action =
        {
          type = "direct",
          action_delivery =
          {
            type = "projectile",
            projectile = "EMP-acid-projectile-big",
            starting_speed = 0.5
          }
        }
    }
end

if data.raw["projectile"]["electric-spit-projectile"] then -- "bob-big-electric-spitter"
	local EMP_acid_splash_electric = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_electric.name = "EMP-acid-splash-electric"
	table.insert(data.raw["projectile"]["electric-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-electric",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_electric})
end

if data.raw["projectile"]["explosive-spit-projectile"] then -- "bob-bigger-spitter" and "bob-big-explosive-worm-turret"
	local EMP_acid_splash_explode = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_explode.name = "EMP-acid-splash-explosive"
	table.insert(data.raw["projectile"]["explosive-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-explosive",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_explode})
end

if data.raw["projectile"]["fire-spit-projectile"] then -- "bob-biggest-spitter" and "bob-big-fire-worm-turret"
	local EMP_acid_splash_fire = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_fire.name = "EMP-acid-splash-fire"
	table.insert(data.raw["projectile"]["fire-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-fire",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_fire})
end

if data.raw["projectile"]["poison-spit-projectile"] then -- "bob-poison-spitter" and "bob-big-poison-worm-turret"
	local EMP_acid_splash_poison = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_poison.name = "EMP-acid-splash-poison"
	table.insert(data.raw["projectile"]["poison-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-poison",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_poison})
end

if data.raw["projectile"]["titan-spit-projectile"] then -- "bob-titan-spitter"
	local EMP_acid_splash_titan = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_titan.name = "EMP-acid-splash-titan"
	table.insert(data.raw["projectile"]["titan-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-titan",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_titan})
end

if data.raw["projectile"]["behemoth-spit-projectile"] then -- "bob-behemoth-spitter"
	local EMP_acid_splash_behemoth = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_behemoth.name = "EMP-acid-splash-behemoth"
	table.insert(data.raw["projectile"]["behemoth-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-behemoth",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_behemoth})
end

if data.raw["projectile"]["leviathan-spit-projectile"] then -- "bob-leviathan-spitter"
	local EMP_acid_splash_leviathan = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_leviathan.name = "EMP-acid-splash-leviathan"
	table.insert(data.raw["projectile"]["leviathan-spit-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-leviathan",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_leviathan})
end

-- Dytech War

if data.raw["projectile"]["berserker-projectile"] then -- All Berzerker-type spitters
	local EMP_acid_splash_berserker = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_berserker.name = "EMP-acid-splash-berserker"
	table.insert(data.raw["projectile"]["berserker-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-berserker",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_berserker})
end

if data.raw["projectile"]["elder-projectile"] then -- All Elder-type spitters
	local EMP_acid_splash_elder = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_elder.name = "EMP-acid-splash-elder"
	table.insert(data.raw["projectile"]["elder-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-elder",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_elder})
end

if data.raw["projectile"]["king-projectile"] then -- All King-type spitters
	local EMP_acid_splash_king = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_king.name = "EMP-acid-splash-king"
	table.insert(data.raw["projectile"]["king-projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-king",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_king})
end

-- Natural Evolution Enemies

if data.raw["projectile"]["Infected-Projectile"] then -- "big-spitter-Mk2"
	local EMP_acid_splash_infected = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_infected.name = "EMP-acid-splash-infected"
	table.insert(data.raw["projectile"]["Infected-Projectile"].action.action_delivery.target_effects, {
		type = "create-entity",
		entity_name = "EMP-acid-splash-infected",
		trigger_created_entity = "true"
	})
	data:extend({EMP_acid_splash_infected})
end

if data.raw["projectile"]["Mutated-Projectile"] then -- "big-spitter-Mk3"
	local EMP_acid_splash_mutated = util.table.deepcopy(data.raw["corpse"]["acid-splash-purple"])
	EMP_acid_splash_mutated.name = "EMP-acid-splash-mutated"
	data.raw["projectile"]["Mutated-Projectile"].action.action_delivery.target_effects[2] = {
		type = "create-entity",
		entity_name = "EMP-acid-splash-mutated",
		trigger_created_entity = "true"
	}
	data:extend({EMP_acid_splash_mutated})
end


control.lua
require "util"
require "config"

script.on_init(function() On_Init() end)
script.on_configuration_changed(function() On_Init() end)
script.on_load(function() On_Load() end)

function On_Init()
	global.tick = global.tick or game.tick
	global.clouds = global.clouds or {}
	if not global.evoFactorFloor then
		if game.forces["enemy"].evolution_factor > 0.995 then
			global.evoFactorFloor = 10
		else
			global.evoFactorFloor = math.floor(game.forces["enemy"].evolution_factor * 10)
		end
		global.tick = global.tick + 1800
	end
end

function On_Load()
	if global.disabledEntities then
		script.on_event(defines.events.on_tick, process_tick)
	end
end

function findEntity(entity)
	for i,t in pairs(global.disabledEntities) do
		if t[1] == entity then
			return i
		end
	end
	return false
end

function entityRemoved(entity)
	local z = findEntity(entity)
	if z then
		local cloud = global.clouds[entity.position.x .. "_" .. entity.position.y]
		if cloud and cloud.valid then
			cloud.destroy()
			global.clouds[entity.position.x .. "_" .. entity.position.y] = nil
		end
		table.remove(global.disabledEntities, z)
	end
end

function process_tick()
	local current_tick = game.tick
	for i = #global.disabledEntities, 1, -1 do -- Loop over table backwards because some entries get removed within the loop
		local entity = global.disabledEntities[i][1]
		if not (entity and entity.valid) then
			table.remove(global.disabledEntities, i)
		elseif global.disabledEntities[i][2] == current_tick then
			entity.active = true
			local cloud = global.clouds[entity.position.x .. "_" .. entity.position.y]
			if cloud and cloud.valid then
				cloud.destroy()
				global.clouds[entity.position.x .. "_" .. entity.position.y] = nil
			end
			table.remove(global.disabledEntities, i)
		else
			if math.random(300) == 7 then
				entity.surface.create_entity{name = "EMP-disabled-sound-" .. math.random(4), position = entity.position}
			end
		end
	end
	if #global.disabledEntities == 0 then
		global.disabledEntities = nil
		script.on_event(defines.events.on_tick, nil)
		global.clouds = {}
	end
end

function disableEntities(entities, min_duration, max_duration)
	local current_tick = game.tick
	for i, entity in pairs(entities) do
		if not global.clouds[entity.position.x .. "_" .. entity.position.y] and entity.health > 0 then
			local cloud = entity.surface.create_entity{name = "disabling-EMP-cloud-" .. math.random(4), position = entity.position}
			global.clouds[entity.position.x .. "_" .. entity.position.y] = cloud
			entity.active = false
			table.insert(global.disabledEntities, {entity, current_tick + math.random(min_duration, max_duration)})
		end
	end
end

function EMPBlast(location, surface, size, min_duration, max_duration)
	if size <= 5 then
		surface.create_entity{name = "EMP-blast-small", position = location}
	elseif size <= 10 then
		surface.create_entity{name = "EMP-blast-medium", position = location}
	elseif size <= 15 then
		surface.create_entity{name = "EMP-blast-large", position = location}
	else
		surface.create_entity{name = "EMP-blast-huge", position = location}
	end
	local EMP_area = {{location.x - size, location.y - size}, {location.x + size, location.y + size}}
	local gun_turrets = surface.find_entities_filtered{area = EMP_area, type="ammo-turret"}
	local fluid_turrets = surface.find_entities_filtered{area = EMP_area, type="fluid-turret"}
	local laser_turrets = surface.find_entities_filtered{area = EMP_area, type="electric-turret"}
	local logistics_bots = surface.find_entities_filtered{area = EMP_area, type="logistic-robot"}
	local construction_bots = surface.find_entities_filtered{area = EMP_area, type="construction-robot"}
	local radars = surface.find_entities_filtered{area = EMP_area, type="radar"}
	if #gun_turrets == 0 and #laser_turrets == 0 and #logistics_bots == 0 and #construction_bots == 0 and #radars == 0 and #fluid_turrets == 0 then
		return
	else
		global.disabledEntities = global.disabledEntities or {}
		disableEntities(gun_turrets, min_duration / 2, max_duration / 2)
		disableEntities(fluid_turrets, min_duration / 2, max_duration / 2)
		disableEntities(laser_turrets, min_duration, max_duration)
		disableEntities(logistics_bots, min_duration, max_duration)
		disableEntities(construction_bots, min_duration, max_duration)
		disableEntities(radars, min_duration, max_duration)
		script.on_event(defines.events.on_tick, process_tick)
	end
end

script.on_event(defines.events.on_entity_died, function(event)
	if global.tick < event.tick then
		if game.forces["enemy"].evolution_factor > 0.995 then
			global.evoFactorFloor = 10
		else
			global.evoFactorFloor = math.floor(game.forces["enemy"].evolution_factor * 10)
		end
		global.tick = global.tick + 1800
	end
	if EMPBiterTable[event.entity.name] then
		local percentChance = EMPBiterTable[event.entity.name][2]
		if math.random() < (global.evoFactorFloor / (1000 / percentChance)) then
			local EMP_size = EMPBiterTable[event.entity.name][1]
			local EMP_min_duration = EMPBiterTable[event.entity.name][3]
			local EMP_max_duration = EMPBiterTable[event.entity.name][4]
			EMPBlast(event.entity.position, event.entity.surface, EMP_size, EMP_min_duration, EMP_max_duration)
		end
	end
	if not global.disabledEntities then
		return
	end
	if vulnerableEntityTypes[event.entity.type] then
		event.entity.active = true
		entityRemoved(event.entity)
	end
end)

script.on_event(defines.events.on_preplayer_mined_item, function(event)
	if not global.disabledEntities then return end
	if vulnerableEntityTypes[event.entity.type] then
		entityRemoved(event.entity)
	end
end)

script.on_event(defines.events.on_robot_pre_mined, function(event)
	if not global.disabledEntities then return end
	if vulnerableEntityTypes[event.entity.type] then
		entityRemoved(event.entity)
	end
end)

script.on_event(defines.events.on_trigger_created_entity, function(event) -- This handles spitter spit
	if EMPSpitterTable[event.entity.name] then
		local percentChance = EMPSpitterTable[event.entity.name][2]
		if math.random() < (global.evoFactorFloor / (1000 / percentChance)) then
			local EMP_size = EMPSpitterTable[event.entity.name][1]
			local EMP_min_duration = EMPSpitterTable[event.entity.name][3]
			local EMP_max_duration = EMPSpitterTable[event.entity.name][4]
			EMPBlast(event.entity.position, event.entity.surface, EMP_size, EMP_min_duration, EMP_max_duration)
		end
	end
end)
